src/main/java/io/thinkit/edc/client/connector/services/ManagementApiHttpClient.java

```diff
@@ -0,0 +1,147 @@
+        return interceptor;
+    }
+
+    protected <T, B> Result<T> send(HttpRequest.Builder requestBuilder, String method, Function<B, T> build) {
+
+        try {
+            var request = interceptor.apply(requestBuilder).build();
+
+            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofInputStream());
+
+            return get(response, method, build);
+        } catch (IOException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected Result<String> send(HttpRequest.Builder requestBuilder, String id) {
+
+        try {
+            var request = interceptor.apply(requestBuilder).build();
+
+            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofInputStream());
+
+            return get(response, id);
+        } catch (IOException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected <T, B> CompletableFuture<Result<T>> sendAsync(
+            HttpRequest.Builder requestBuilder, String method, Function<B, T> build) {
+
+        var request = interceptor.apply(requestBuilder).build();
+
+        CompletableFuture<HttpResponse<InputStream>> future =
+                httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofInputStream());
+        return future.thenApply(response -> get(response, method, build));
+    }
+
+    protected CompletableFuture<Result<String>> sendAsync(HttpRequest.Builder requestBuilder, String id) {
+
+        var request = interceptor.apply(requestBuilder).build();
+
+        CompletableFuture<HttpResponse<InputStream>> future =
+                httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofInputStream());
+        return future.thenApply(response -> get(response, id));
+    }
+
+    protected Result<String> get(HttpResponse<InputStream> response, String id) {
+        try {
+            var statusCode = response.statusCode();
+            if (isSuccessful(statusCode)) {
+                if (id.isEmpty()) {
+                    var content = expand(response.body());
+                    var result = content.getJsonObject(0).getString(ID);
+                    return new Result<>(result, null);
+                }
+                return new Result<>(id, null);
+
+            } else {
+                var error = deserializeToArray(response.body()).stream()
+                        .map(s -> new ApiErrorDetail(s.asJsonObject()))
+                        .toList();
+                return new Result<>(error);
+            }
+        } catch (JsonLdError e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected <T, B> Result<T> get(HttpResponse<InputStream> response, String method, Function<B, T> build) {
+        try {
+            var statusCode = response.statusCode();
+            if (isSuccessful(statusCode)) {
+                T result = null;
+                switch (method) {
+                    case "get" -> {
```

Comment:
I think this `method` thing could be improved, because relying on a string is not the safest thing to do, and it makes the fully coupled and impossible to extend (e.g. if any controller would need a special way to deserialize the body there will be the need to modify the client adding a new switch branch).

Instead of passing a string that determines which function that maps from the `InputStream` to the `JsonObject`/`JsonArray`, you could have a deserialization function, that has a signature like:
`Function<InputStream, B>`, with `B` that is the type that then will be input of the `build` function (that will need a better name anyway), so also the casts at line 118 and 124 will be unnecessary because it will be already handled by the type boundaries.

The two functions could be called `get` and `getStatus` as you did, but maybe they could be called `toJsonLd` and `toSimpleJson` and will be exposed by the `ManagementApiHttpClient` class so they can be used by the controllers directly.

This refactoring will put full control in hands of the controllers, that eventually (not needed in this case) could implement a custom logic for deserializing a different kind of body.